modules/livestore/instance.go

```diff
@@ -0,0 +1,500 @@
+
+	// Create completed block
+	reader := backend.NewReader(i.wal.LocalBackend())
+	writer := backend.NewWriter(i.wal.LocalBackend())
+
+	iter, err := walBlock.Iterator()
+	if err != nil {
+		level.Error(i.logger).Log("msg", "failed to get WAL block iterator", "id", id, "err", err)
+		span.RecordError(err)
+		return err
+	}
+	defer iter.Close()
+
+	newMeta, err := i.enc.CreateBlock(ctx, &i.Cfg.BlockConfig, walBlock.BlockMeta(), iter, reader, writer)
+	if err != nil {
+		level.Error(i.logger).Log("msg", "failed to create complete block", "id", id, "err", err)
+		span.RecordError(err)
+		return err
+	}
+
+	newBlock, err := i.enc.OpenBlock(newMeta, reader)
+	if err != nil {
+		level.Error(i.logger).Log("msg", "failed to open complete block", "id", id, "err", err)
+		span.RecordError(err)
+		return err
+	}
+
+	i.blocksMtx.Lock()
+	// Verify the WAL block still exists
+	if _, ok := i.walBlocks[id]; !ok {
+		level.Warn(i.logger).Log("msg", "WAL block disappeared while being completed, deleting complete block", "id", id)
+		err := i.wal.LocalBackend().ClearBlock(id, i.tenantID)
+		if err != nil {
+			level.Error(i.logger).Log("msg", "failed to clear complete block after WAL disappeared", "block", id, "err", err)
+		}
+		i.blocksMtx.Unlock()
+		span.AddEvent("WAL block disappeared during completion")
+		return nil
+	}
+
+	i.completeBlocks[id] = ingester.NewLocalBlock(ctx, newBlock, i.wal.LocalBackend())
+
+	err = walBlock.Clear()
+	if err != nil {
+		level.Error(i.logger).Log("msg", "failed to clear WAL block", "id", id, "err", err)
+		span.RecordError(err)
+	}
+	delete(i.walBlocks, (uuid.UUID)(walBlock.BlockMeta().BlockID))
+	i.blocksMtx.Unlock()
+
+	level.Info(i.logger).Log("msg", "completed block", "id", id.String())
+	span.AddEvent("block completed successfully")
+	return nil
+}
+
+func (i *instance) deleteOldBlocks() error {
+	i.blocksMtx.Lock()
+	defer i.blocksMtx.Unlock()
+
+	cutoff := time.Now().Add(-i.Cfg.CompleteBlockTimeout) // Delete blocks older than Complete Block Timeout
+
+	for id, walBlock := range i.walBlocks {
+		if walBlock.BlockMeta().EndTime.Before(cutoff) {
+			if _, ok := i.completeBlocks[id]; !ok {
+				level.Warn(i.logger).Log("msg", "deleting WAL block that was never completed", "block", id.String())
+			}
+			err := walBlock.Clear()
+			if err != nil {
+				return err
+			}
+			delete(i.walBlocks, id)
+			metricBlocksClearedTotal.WithLabelValues("wal").Inc()
+		}
+	}
+
+	for id, completeBlock := range i.completeBlocks {
+		if completeBlock.BlockMeta().EndTime.Before(cutoff) {
+
+			level.Info(i.logger).Log("msg", "deleting complete block", "block", id.String())
+			err := i.wal.LocalBackend().ClearBlock(id, i.tenantID)
+			if err != nil {
+				return err
+			}
+			delete(i.completeBlocks, id)
+			metricBlocksClearedTotal.WithLabelValues("complete").Inc()
+		}
+	}
+
+	return nil
+}
+
+func (i *instance) FindByTraceID(ctx context.Context, traceID []byte) (*tempopb.TraceByIDResponse, error) {
+	i.blocksMtx.RLock()
```

Comment:
Can this be moved down to after live traces, is there a need to lock both at the same time?